### PR TITLE
Added following space option into autocomplete

### LINKED ISSUE #2008: Autocomplete should put the selection after output template
## Are you reporting a feature request or a bug?

Bug

## Provide detailed reproduction steps (if any)

This plugin is based on not yet published Mentions feature branch `t/1703-b`.

1. Open [output template manual test](http://tests.ckeditor.test:1030/tests/plugins/autocomplete/manual/outputtemplate).
1. Focus the editor.
1. Type "@" to start autocompleteion.
1. Accept first entry.
1. Type " and " (note spaces).
1. Type "@" to start another autocompletion.
1. Accept first entry.

### Expected result

Text between the autocompleted values should not be bolded.

### Actual result

The text is bolded.

Reason is that the selection gets inside of the inserted element.

There are two solutinos to that:

* Put a selection outside that element (might be impossible on WebKits due to their normalization).
* Add an option to insert a space after the insertion. This option could be technically realized by adding a space to `outputTemplate` but the problem is if there is already a space after the text where you started autocompletion, it would just stupidly insert a space there resulting with two spaces.

## Other details

This issue is a follow up of https://github.com/ckeditor/ckeditor-dev/pull/1992#pullrequestreview-123272523.

* Browser: Any
* OS: Any
* CKEditor version: 4.10.0

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

The following space can be added at the end of the inserted word. This feature also recognizes existing space.

Closes #2008

### CODE REVIEW

**Comment by jacekbogdanski** [comment]
Rebased into the latest upstream.

File: tests/plugins/autocomplete/manual/followingspace.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+@bender-tags: 4.10.0, feature, 2008
```
Comment (msamsel) [inline tests/plugins/autocomplete/manual/followingspace.md]: Tag needs to be updated.

File: tests/plugins/autocomplete/autocomplete.js [hunk 2]
```diff
@@ -563,6 +565,57 @@
...
+				ac = new CKEDITOR.plugins.autocomplete( editor, {
+					dataCallback: dataCallback,
+					textTestCallback: textTestCallback
+				} );
+
+			this.editorBots.standard.setHtmlWithSelection( '' );
+
+			editor.editable().fire( 'keyup', new CKEDITOR.dom.event( {} ) );
+
+			editable.fire( 'keydown', new CKEDITOR.dom.event( { keyCode: 13 } ) ); // ENTER
```
[the remark this answers is not in the document]
Comment (msamsel) [reply tests/plugins/autocomplete/autocomplete.js]: Comment is for 2 lines higher line. I missclicked it.

**Comment by jacekbogdanski** [comment]
Your discovery makes me wonder if we should enable `followingSpace` by default. This PR was aiming to autocomplete release previously, but currently, we will change autocomplete and mentions behavior which may break users output templates.

Because of that, I disabled `followingSpace` by default.

File: plugins/mentions/plugin.js [hunk 3]
```diff
@@ -106,6 +106,13 @@
 		 */
 		this.cache = config.cache !== undefined ? config.cache : true;
 
+		/**
+		 * @inheritdoc CKEDITOR.plugins.autocomplete#followingSpace
+		 * @property {Boolean} [followingSpace]
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/mentions/plugin.js]: Actually, there is a default value described by brackets `[ ... ]` which means that this property is optional -> it's `false`.

File: tests/plugins/autocomplete/manual/followingspace.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+					textTestCallback: autocompleteUtils.getTextTestCallback(),
+					dataCallback: autocompleteUtils.getDataCallback(),
+					followingSpace: true
+				} );
+			}
+		}
+	};
+
+	CKEDITOR.replace( 'classic', config );
+	CKEDITOR.inline( 'inline', config );
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/autocomplete/manual/followingspace.html]: https://github.com/ckeditor/ckeditor-dev/pull/2078#discussion_r222292789

**Comment by jacekbogdanski** [comment]
It seems like it's easier to just remove doubled space than inserting a new one. The reason is that we have to move with selection outside output template, although it should be still placed inside completion range to keep preselected styles.

Unit tests are already written e.g. `test following space with output template`. I also added an additional manual test for the issue.

File: plugins/autocomplete/plugin.js [hunk 5]
```diff
@@ -1344,6 +1362,20 @@
 		}, {} );
 	}
 
+	function removeLeadingSpace( editor ) {
+		var selection = editor.getSelection(),
+			nextNode = selection.getRanges()[ 0 ].getNextNode( function( node ) {
+			return Boolean( node.type == CKEDITOR.NODE_TEXT && node.getText() );
```
Comment (msamsel) [inline plugins/autocomplete/plugin.js]: This and next line seems to have wrong indentions.

File: plugins/autocomplete/plugin.js [hunk 6]
```diff
@@ -1344,6 +1362,20 @@
...
+		} );
+
+		if ( nextNode && nextNode.getText().match( /^\s+/ ) ) {
+			var range = editor.createRange();
```
Comment (msamsel) [inline plugins/autocomplete/plugin.js]: There should be only one var statement at the beginning of the function.

File: plugins/mentions/plugin.js [hunk 7]
```diff
@@ -106,6 +106,13 @@
 		 */
 		this.cache = config.cache !== undefined ? config.cache : true;
 
+		/**
```
Comment (msamsel) [inline plugins/mentions/plugin.js]: There is missing `@since` tag, which is not inherited.

File: plugins/autocomplete/plugin.js (hunk 6 above)
Comment (jacekbogdanski) [reply plugins/autocomplete/plugin.js]: `var` statement in most cases is the most readable when it's as close as possible to the first usage - as it's a subjective matter, it still objective that in the above use case you will unnecessary create a range object if `if` statement fail.

**Comment by jacekbogdanski** [comment]
> I see that you implement new configuration option for mention plugin, however there is no test which covers that change. I know it just pass vale to autocomplete, but in case that autocomplete will introduce some bug, then mention plugin might have chance to detect it.

Mentions behaves as a wrapper around autocomplete and with such approach we have to duplicate every test for mentions which is placed in autocomplete. It probably would be the best solution, however it creates a lot of code to maintain. I propose to keep 1:1 ratio between tests and features which is done in most our code base and also in mentions plugin (you can see how it has been done for `throttle` feature).

However, I think it would be nice to test if the values are really passed to autocomplete against breaking changes in constructor. I already added unit test for it.

File: plugins/autocomplete/plugin.js (hunk 6 above)
Comment (msamsel) [reply plugins/autocomplete/plugin.js]: You don't have to assign value with declaration of this variable, so range don't have to be created in such case and you just assign proper value inside `if` statement.
As it's quite small function and there is hoisting, from my perspective it's better to declare it at the beginning of scope where it used. However I understand your concerns and let remain it for main review to decide wether current approach should be changed or not.

File: tests/plugins/mentions/mentions.js [hunk 8]
```diff
@@ -462,6 +462,30 @@
...
+
+			this.createMentionsInstance( config );
+
+			spy.restore();
+
+			assert.isTrue( spy.calledWith( this.editor, sinon.match( compareConfig ) ) );
+
+			function compareConfig( val ) {
+				for ( var k in config ) {
+					assert.areEqual( config[ k ], val[ k ], 'Option: ' + k + ' should be passed to autocomplete.' );
```
Comment (msamsel) [inline tests/plugins/mentions/mentions.js]: You can use `objectAssert.areEqual` or `objectAssert.areDeepEqual` methods to compare objects.

**Comment by jacekbogdanski** [comment]
Actually, I can't, these two config objects are not equal. Autocomplete config contains also callbacks created internally inside mentions plugin.

Additionally I updated config validation with template options: https://github.com/ckeditor/ckeditor-dev/pull/2078/commits/3f8f90ed028b3db6584545766ccbfdb9f52a50f1

File: tests/plugins/mentions/mentions.js (hunk 8 above)
Comment (msamsel) [reply tests/plugins/mentions/mentions.js]: https://github.com/ckeditor/ckeditor-dev/pull/2078#issuecomment-448494237

**Pull request closed without merging**